operator/controllers/btpoperator_controller.go

```diff
@@ -347,3 +417,283 @@ func (r *BtpOperatorReconciler) watchBtpOperatorUpdatePredicate() predicate.Func
 		},
 	}
 }
+
+func (r *BtpOperatorReconciler) handleDeprovisioning(ctx context.Context) error {
+	logger := log.FromContext(ctx)
+	logger.Info("btp operator is under deletion")
+
+	namespaces := &corev1.NamespaceList{}
+	if err := r.List(ctx, namespaces); err != nil {
+		return err
+	}
+	if err := r.handlePreDelete(ctx); err != nil {
+		return err
+	}
+
+	hardDeleteChannel := make(chan bool)
+	go r.handleHardDelete(ctx, namespaces, hardDeleteChannel)
+
+	select {
+	case hardDeleteOk := <-hardDeleteChannel:
+		if hardDeleteOk {
+			logger.Info("hard delete success")
+			if err := r.cleanUpAllBtpOperatorResources(ctx, namespaces); err != nil {
+				logger.Error(err, "failed to remove related installed resources")
+				return err
+			}
+		} else {
+			if err := r.handleSoftDelete(ctx, namespaces); err != nil {
+				return err
+			}
+		}
+	case <-time.After(r.reconcilerConfig.timeout):
+		if err := r.handleSoftDelete(ctx, namespaces); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *BtpOperatorReconciler) handleHardDeleteTest(anyErr *bool) {
+	if r.reconcilerConfig.testFlag == testScenarioWithError {
+		*anyErr = true
+	}
+	if r.reconcilerConfig.testFlag == testScenarioWithTimeout {
+		time.Sleep(time.Second * 5)
+	}
+}
+
+func (r *BtpOperatorReconciler) handleHardDelete(ctx context.Context, namespaces *corev1.NamespaceList, success chan bool) {
+	defer close(success)
+	logger := log.FromContext(ctx)
+
+	anyErr := false
+	if err := r.hardDelete(ctx, bindingGvk, namespaces); err != nil {
+		logger.Error(err, "hard delete of binding failed")
+		anyErr = true
+	}
+	if err := r.hardDelete(ctx, instanceGvk, namespaces); err != nil {
+		logger.Error(err, "hard delete of instance failed")
+		anyErr = true
+	}
+
+	if r.reconcilerConfig.testFlag != "" {
+		r.handleHardDeleteTest(&anyErr)
+	}
+
+	if anyErr {
+		success <- false
+		return
+	}
+
+	for {
+		err, resourcesLeft := r.checkIfAnyResourcesLeft(ctx, namespaces)
+		if err != nil {
+			logger.Error(err, "Checking failed")
+			success <- false
+			return
+		}
+		if !resourcesLeft {
+			success <- true
+			return
+		}
+		time.Sleep(retryInterval)
```

Comment:
this can potentially lead to endless loop

Reply:
main thread will continue after the timeout, but handleHardDelete() runs in a separate goroutine which can be in the loop, example:
https://go.dev/play/p/Qfm8mmX9j3H